### PR TITLE
Use a single implementation for `combine_table_entry`

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR generalises `JointLookup` so that it can handle `SingleLookup<F>`s, `F`s, or `Expr<ConstantExpr<F>>`s as arguments, and relaxes the constraints on `combine_table_entry`.

This allows us to
* store the 'dummy' lookup value as a `JointLookup` in the constraint system, rather than unpacking and re-packing it
  - not done here; this will be in a follow-up PR
* unify the 3 different implementations of `combine_table_entry` into one
  - happily, the 'fold' implementation in `combine_table_entry` minimises the number of multiplications compared to the others, so this is also an efficiency gain
* have a single point to change when we add `table_id`s

### CODE REVIEW

File: kimchi/src/circuits/expr.rs [hunk 1]
```diff
@@ -1921,6 +1921,14 @@ impl<F: Field> From<u64> for Expr<ConstantExpr<F>> {
     }
 }
 
+impl<F: Field> Mul<F> for Expr<ConstantExpr<F>> {
+    type Output = Expr<ConstantExpr<F>>;
+
+    fn mul(self, y: F) -> Self::Output {
+        Expr::Constant(ConstantExpr::Literal(F::from(y))) * self
```
Comment (mimoo) [inline kimchi/src/circuits/expr.rs:1928]: you might want to either remove the `F::from` or have `F: Into<Field>`

File: kimchi/src/circuits/gate.rs [hunk 2]
```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
...
+    'a,
+    F: 'a + Zero + One + Add + Mul + Clone,
+    I: DoubleEndedIterator<Item = &'a F>,
+>(
     joint_combiner: F,
     v: I,
 ) -> F {
-    v.rev().fold(F::zero(), |acc, x| joint_combiner * acc + x)
+    v.rev()
+        .fold(F::zero(), |acc, x| joint_combiner.clone() * acc + x.clone())
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:81]: if you add Copy instead of Clone you can remove the clone here (Field is Copy)

Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:81]: `F` is copy but `Expr` isn't (and we use `JointCombiner` rather than a field element there).

File: kimchi/src/circuits/gate.rs [hunk 3]
```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
 /// analogously using `joint_combiner`.
 ///
 /// This function computes that combined value.
-pub fn combine_table_entry<'a, F: Field, I: DoubleEndedIterator<Item = &'a F>>(
+pub fn combine_table_entry<
+    'a,
+    F: 'a + Zero + One + Add + Mul + Clone,
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:74]: I'm wondering why `F: Field + Zero` here is not enough?

Also I'm wondering how I'm supposed to read `F: 'a`? And why it doesn't work without that

File: kimchi/src/circuits/gate.rs [hunk 4]
```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
...
-    v.rev().fold(F::zero(), |acc, x| joint_combiner * acc + x)
+    v.rev()
+        .fold(F::zero(), |acc, x| joint_combiner.clone() * acc + x.clone())
 }
 
-impl<F: Field> SingleLookup<F> {
+impl<F: Copy> SingleLookup<F> {
     /// Evaluate the linear combination specifying the lookup value to a field element.
-    pub fn evaluate<G: Fn(LocalPosition) -> F>(&self, eval: G) -> F {
+    pub fn evaluate<K: Zero + Add + Mul<F, Output = K>, G: Fn(LocalPosition) -> K>(
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:86]: in general it's nicer to move these bounds to a `where` clause at the end of the function signature

File: kimchi/src/circuits/gate.rs [hunk 5]
```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
...
     joint_combiner: F,
     v: I,
 ) -> F {
-    v.rev().fold(F::zero(), |acc, x| joint_combiner * acc + x)
+    v.rev()
+        .fold(F::zero(), |acc, x| joint_combiner.clone() * acc + x.clone())
 }
 
-impl<F: Field> SingleLookup<F> {
+impl<F: Copy> SingleLookup<F> {
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:84]: is there a reason behind that relaxation and the creation of K? Why not just take everything as a `Field`?

File: kimchi/src/circuits/gate.rs [hunk 6]
```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
...
 }
 
 /// A spec for checking that the given vector belongs to a vector-valued lookup table.
 #[derive(Clone, Serialize, Deserialize)]
-pub struct JointLookup<F> {
-    pub table_id: usize,
-    pub entry: Vec<SingleLookup<F>>,
+pub struct JointLookup<SingleLookup> {
+    pub table_id: i32,
+    pub entry: Vec<SingleLookup>,
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:100]: why not just make this `Vec<SingleLookup<F>>` and the struct generic in `F`?

File: kimchi/src/circuits/gate.rs [hunk 7]
```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
...
-        for s in self.entry.iter() {
-            res += c * s.evaluate(eval);
-            c *= joint_combiner;
+    pub fn evaluate(&self, joint_combiner: F) -> F {
+        combine_table_entry(joint_combiner, self.entry.iter())
+    }
+}
+
+impl<F: Copy> JointLookup<SingleLookup<F>> {
+    pub fn reduce<K: Zero + Mul<F, Output = K>, G: Fn(LocalPosition) -> K>(
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:114]: same here

File: kimchi/src/circuits/gate.rs (hunk 6 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:100]: Because `reduce` erases `SingleLookup` and returns a concrete value/`Expr` for the lookup

File: kimchi/src/circuits/gate.rs (hunk 5 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:84]: `K` can be `Expr`; that was the point of this PR.

File: kimchi/src/circuits/gate.rs (hunk 2 above)
Comment (mimoo) [reply kimchi/src/circuits/gate.rs:81]: ah I see. My tolerance for generics is usually pretty low ^^ I'm wondering how we can make this clearer. Maybe by changing `F` with `F_or_Expr` or something. Food for thought

Comment (mimoo) [reply kimchi/src/circuits/gate.rs:81]: and then creating an actual trait `F_or_Expr : Zero + One + ...` and implement it on what we need

File: kimchi/src/circuits/gate.rs (hunk 5 above)
Comment (mimoo) [reply kimchi/src/circuits/gate.rs:84]: *facepalm* I did get straight into the code without reading the description

File: kimchi/src/circuits/gate.rs [hunk 8]
```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
...
+        &self,
+        eval: &G,
+    ) -> JointLookup<K> {
+        JointLookup {
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:118]: can you Self in place of JointLookup here. I'm not sure it's objectively better actually... but I've been conditioned to do this

File: kimchi/src/circuits/gate.rs [hunk 9]
```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
...
+            table_id: self.table_id,
+            entry: self.entry.iter().map(|s| s.evaluate(eval)).collect(),
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:120]: I think this should work too:

```rust
let entry = self.entry.iter().map(|s| s.evaluate(eval)).collect();
Self { entry, ..self }
```

File: kimchi/src/circuits/gate.rs (hunk 6 above)
Comment (mimoo) [reply kimchi/src/circuits/gate.rs:100]: yeah I see that now. So it's like we have multiple types. Do you think it'd be a good idea to define them somewhere and use them? Something like that:

```rust
type EvaluatedJointLookup<F> = JointLookup<F>;
type ExprSingleLookup<F> = SingleLookup<ConstantExpr<F>>;
type ExprJointLookup<F> = JointLookup<SingleLookup<ConstantExpr<F>>>;
```

might be a stupid idea

Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:100]: We never use any of those types directly, not sure it's worth it.

File: kimchi/src/circuits/gate.rs (hunk 3 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:74]: `F` may be a type that is defined within a function; if the scope of `'a` is broader than the function, then the data outlives the type. Thus `F: 'a`.

File: kimchi/src/circuits/polynomials/lookup.rs [hunk 10]
```diff
@@ -661,16 +635,26 @@ pub fn constraints<F: FftField>(configuration: &LookupConfiguration<F>, d1: D<F>
             .collect()
     };
 
+    let eval = |pos: LocalPosition| {
+        E::Cell(Variable {
+            col: Column::Witness(pos.column),
+            row: pos.row,
+        })
+    };
```
Comment (mimoo) [inline kimchi/src/circuits/polynomials/lookup.rs:643]: we already have these helpers in expr.rs

File: kimchi/src/circuits/polynomials/lookup.rs [hunk 11]
```diff
@@ -661,16 +635,26 @@ pub fn constraints<F: FftField>(configuration: &LookupConfiguration<F>, d1: D<F>
...
-    let f_term = |spec: &Vec<_>| {
+    let f_term = |spec: &Vec<JointLookupSpec<_>>| {
         assert!(spec.len() <= lookup_info.max_per_row);
         let padding = complements_with_beta_term[lookup_info.max_per_row - spec.len()].clone();
 
         spec.iter()
-            .map(|j| E::Constant(ConstantExpr::Gamma) + joint_lookup(j))
+            .map(|j| {
+                E::Constant(ConstantExpr::Gamma)
+                    + j.evaluate(E::constant(ConstantExpr::JointCombiner), &eval)
```
Comment (mimoo) [inline kimchi/src/circuits/polynomials/lookup.rs:656]: I'm confused as to how you can pass a reference to a closure to evaluate, which doesn't seem to expect a closure there

Comment (mimoo) [reply kimchi/src/circuits/polynomials/lookup.rs:656]: I think rust-analyzer is getting confused as to which `evaluate()` function is called o_O

Comment (mimoo) [reply kimchi/src/circuits/polynomials/lookup.rs:656]: https://github.com/rust-analyzer/rust-analyzer/issues/11759

File: kimchi/src/circuits/gate.rs [hunk 12]
```diff
@@ -68,40 +69,67 @@ pub struct SingleLookup<F> {
...
+
+    pub fn evaluate<
+        K: Zero + One + Add + Mul + Mul<F, Output = K> + Clone,
+        G: Fn(LocalPosition) -> K,
+    >(
+        &self,
+        joint_combiner: K,
+        eval: &G,
+    ) -> K {
+        self.reduce(eval).evaluate(joint_combiner)
```
Comment (mimoo) [inline kimchi/src/circuits/gate.rs:132]: honestly I find this function really hard to parse.

* we have a `JointLookup<SingleLookup<F>>` for some `F: Zero + One + Add + Mul + Clone`
* we want a `K` for some `K: Zero + One + Add + Mul + Mul<F, Output = K> + Clone`
* in the mean time, we have some eval function that takes a `LocalPosition` and return a `K`

it's just too generic to really grasp what is really going on IMO. Is there a way to make this clearer?

Comment (mimoo) [reply kimchi/src/circuits/gate.rs:132]: interestingly my IDE is also working against myself, pointing me to the wrong reduce function

Comment (mimoo) [reply kimchi/src/circuits/gate.rs:132]: I'm wondering if we can just have code duplication for the different instantiations of these functions, this would make the code much clearer imo and we wouldn't need to pass a closure around (IIUC)

File: kimchi/src/circuits/gate.rs (hunk 9 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:120]: Types mismatch, `self: JointLookup<SingleLookup<F>>`, return value is `JointLookup<F>`.

File: kimchi/src/circuits/polynomials/lookup.rs (hunk 11 above)
Comment (mrmr1993) [reply kimchi/src/circuits/polynomials/lookup.rs:656]: Commented there; presumably it's due to name reuse.

File: kimchi/src/circuits/gate.rs (hunk 8 above)
Comment (mrmr1993) [reply kimchi/src/circuits/gate.rs:122]: I don't think I can, because the return type is different from `Self`..

File: kimchi/src/circuits/polynomials/lookup.rs [hunk 13]
```diff
@@ -142,29 +139,6 @@ use CurrOrNext::*;
...
-                })
-        })
-        .fold(E::zero(), |acc, e| acc + e)
-}
-
-fn joint_lookup<F: FftField>(j: &JointLookup<F>) -> E<F> {
-    j.entry
-        .iter()
-        .enumerate()
-        .map(|(i, s)| E::constant(ConstantExpr::JointCombiner.pow(i as u64)) * single_lookup(s))
```
Comment (mimoo) [inline kimchi/src/circuits/polynomials/lookup.rs:164]: so if I understand correctly this is replaced by the logic in `combine_table_entry`

**Review by mimoo — APPROVED** [review]
Approved, but I find the generic implementation really hard to follow, I'm even wondering how I'll ever be able to change this code without keeping 10 different low-level details in my head.

File: kimchi/src/circuits/polynomials/lookup.rs (hunk 13 above)
Comment (mrmr1993) [reply kimchi/src/circuits/polynomials/lookup.rs:164]: Exactly, yes
